### PR TITLE
Error info has been added to LoadState::Failed

### LINKED ISSUE #12667: bevy_asset `LoadState` should contain more useful information in its `Failed` arm
## What problem does this solve or what need does it fill?

When checking if the required asset are loaded, I'll call `asset_server.get_load_state(handle)` to check if my assets are loaded yet.

If it fails, I would like to report a detailed error to the user / developer, so they can troubleshoot the problem. Instead, I simply get `LoadState::Failed`, with no information.

## What solution would you like?

Store [`AssetLoadError`](https://docs.rs/bevy/latest/bevy/asset/enum.AssetLoadError.html) inside of  [`LoadState::Failed`](https://docs.rs/bevy/latest/bevy/asset/enum.LoadState.html). This may be wise to store as a `Box<AssetLoadError>` to avoid bloating the size of all variants of `LoadState`.

## What alternative(s) have you considered?

I could listen to the `AssetLoadFailedEvent<A: Asset>`, and log those as they occurred. This is not as convenient (or easy to correlate), and won't work for untyped asset loading solutions.

### PR DESCRIPTION
**Pull request by bugsweeper**
# Objective

Fixes #12667.

## Solution

- Stored [AssetLoadError](https://docs.rs/bevy/latest/bevy/asset/enum.AssetLoadError.html) inside of [LoadState::Failed](https://docs.rs/bevy/latest/bevy/asset/enum.LoadState.html) as a Box<AssetLoadError> to avoid bloating the size of all variants of LoadState.
- Fixed dependent code

## Migration guide

Added [AssetLoadError](https://docs.rs/bevy/latest/bevy/asset/enum.AssetLoadError.html) to [LoadState::Failed](https://docs.rs/bevy/latest/bevy/asset/enum.LoadState.html) option
Removed `Copy`, `Ord` and `PartialOrd` implementations for [LoadState](https://docs.rs/bevy/latest/bevy/asset/enum.LoadState.html) enum
Added `Eq` and `PartialEq` implementations for [MissingAssetSourceError](https://docs.rs/bevy/latest/bevy/asset/io/struct.MissingAssetSourceError.html), [MissingProcessedAssetReaderError](https://docs.rs/bevy/latest/bevy/asset/io/struct.MissingProcessedAssetReaderError.html), [DeserializeMetaError](https://docs.rs/bevy/latest/bevy/asset/enum.DeserializeMetaError.html), [LoadState](https://docs.rs/bevy/latest/bevy/asset/enum.LoadState.html), [AssetLoadError](https://docs.rs/bevy/latest/bevy/asset/enum.AssetLoadError.html), [MissingAssetLoaderForTypeNameError](https://docs.rs/bevy/latest/bevy/asset/struct.MissingAssetLoaderForTypeNameError.html) and [MissingAssetLoaderForTypeIdError](https://docs.rs/bevy/latest/bevy/asset/struct.MissingAssetLoaderForTypeIdError.html)

### CODE REVIEW

File: crates/bevy_asset/src/server/mod.rs [hunk 1]
```diff
@@ -1199,9 +1203,18 @@ pub enum LoadState {
     /// The asset has been loaded and has been added to the [`World`]
     Loaded,
     /// The asset failed to load.
-    Failed,
+    Failed(Box<AssetLoadError>),
 }
 
+impl PartialEq for LoadState {
```
Comment (alice-i-cecile) [inline crates/bevy_asset/src/server/mod.rs:1209]: I don't think we want this: IMO we want to use the standard derived equality here. If users want this behavior it's easy for them to use it manually.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Really pleased to see this! I think we should stick with the derived `PartialEq` and `Eq` implementations (and derive down the tree as needed), but otherwise I'm pleased.

We should also derive `Debug` and `Error` (via `thiserror`) while we're here.

File: crates/bevy_asset/src/server/mod.rs (hunk 1 above)
Comment (bugsweeper) [reply crates/bevy_asset/src/server/mod.rs:1209]: Standard derived equality requires recursive equality implementations for all nested structures, should I add derived equality for all of them?

Comment (alice-i-cecile) [reply crates/bevy_asset/src/server/mod.rs:1209]: Yes please :) Being able to check that the error matches the reason you expect is really valuable for writing tests, both internally and externally.

**Comment by bugsweeper** [comment]
Looks like failed checks related to #12205 , but  my code doesn't have relation to it, how can I resolve problem @alice-i-cecile ?

**Comment by bugsweeper** [comment]
@alice-i-cecile I made [some changes](https://github.com/bevyengine/bevy/pull/12709#commits-pushed-f2e4101) after your review. And now all checks have passed. I hope you will find time for a feedback.

File: crates/bevy_asset/src/server/mod.rs [hunk 2]
```diff
@@ -1286,22 +1287,47 @@ pub enum AssetLoadError {
...
+#[error("Failed to load asset '{path}' with asset loader '{loader_name}': {error}")]
+pub struct AssetLoaderError {
+    path: AssetPath<'static>,
+    loader_name: &'static str,
+    error: Arc<dyn std::error::Error + Send + Sync + 'static>,
+}
+
+impl PartialEq for AssetLoaderError {
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
```
Comment (alice-i-cecile) [inline crates/bevy_asset/src/server/mod.rs:1300]: We should call out in a doc string for this implementation that we're only comparing the `TypeId` of the underlying error.

This is somewhat surprising, so we should mention it to save people time when debugging.

File: crates/bevy_asset/src/server/mod.rs [hunk 3]
```diff
@@ -1190,16 +1194,17 @@ pub(crate) enum InternalAssetEvent {
 }
 
 /// The load state of an asset.
-#[derive(Component, Copy, Clone, Debug, Eq, PartialEq, Ord, PartialOrd)]
+#[derive(Error, Component, Clone, Debug, PartialEq, Eq)]
```
Comment (alice-i-cecile) [inline crates/bevy_asset/src/server/mod.rs:1197]: This type shouldn't impl `Error`: it's analgous to a `Result`. Sorry for any miscommunication.

**Comment by alice-i-cecile** [comment]
Two small notes, but we're almost there :)

**Comment by bugsweeper** [comment]
@alice-i-cecile this PR is ready for your review

**Review by alice-i-cecile — APPROVED** [review]
Thank you very much!

File: crates/bevy_asset/src/server/mod.rs (hunk 2 above)
Comment (cart) [reply crates/bevy_asset/src/server/mod.rs:1300]: Probably want to use different phrasing than "partial" here:

> is partial (only through `TypeId`)

The "partial" in PartialEq refers to non-reflexiveness, not "partial comparison".

File: crates/bevy_asset/src/server/mod.rs [hunk 4]
```diff
@@ -1286,22 +1286,48 @@ pub enum AssetLoadError {
...
 #[derive(Error, Debug, Clone)]
+#[error("Failed to load asset '{path}' with asset loader '{loader_name}': {error}")]
+pub struct AssetLoaderError {
+    path: AssetPath<'static>,
+    loader_name: &'static str,
+    error: Arc<dyn std::error::Error + Send + Sync + 'static>,
+}
+
+impl PartialEq for AssetLoaderError {
+    /// Equality comparison for `AssetLoaderError::error` is partial (only through `TypeId`)
```
Comment (cart) [inline crates/bevy_asset/src/server/mod.rs:1298]: Probably want to use different phrasing than "partial" here (and in the other location:

>   is partial (only through TypeId)

The "partial" in PartialEq refers to non-reflexiveness, not "partial comparison"

**Review by cart — APPROVED** [review]
I'm on board for this!

File: crates/bevy_asset/src/server/mod.rs (hunk 4 above)
Comment (bugsweeper) [reply crates/bevy_asset/src/server/mod.rs]: replaced "partial" to "not full"

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1297 if you'd like to help out.

**Pull request merged**